eclair-core/src/main/scala/fr/acinq/eclair/router/Validation.scala

```diff
@@ -414,51 +415,63 @@ object Validation {
-        // when we're sending updates to ourselves
-        origins.head match {
-          case RemoteGossip(peerConnection, remoteNodeId) =>
-            val query = QueryShortChannelIds(u.chainHash, EncodedShortChannelIds(routerConf.encodingType, List(realShortChannelId)), TlvStream.empty)
-            d.sync.get(remoteNodeId) match {
-              case Some(sync) if sync.started =>
-                // we already have a pending request to that node, let's add this channel to the list and we'll get it later
-                d.copy(sync = d.sync + (remoteNodeId -> sync.copy(remainingQueries = sync.remainingQueries :+ query, totalQueries = sync.totalQueries + 1)))
-              case _ =>
-                // otherwise we send the query right away
-                peerConnection ! query
-                d.copy(sync = d.sync + (remoteNodeId -> Syncing(remainingQueries = Nil, totalQueries = 1)))
-            }
-          case _ =>
-            // we don't know which node this update came from (maybe it was stashed and the channel got pruned in the meantime or some other corner case).
-            // or we don't have a peerConnection to send our query to.
-            // anyway, that's not really a big deal because we have removed the channel from the pruned db so next time it shows up we will revalidate it
-            d
+          d.prunedChannels.get(RealShortChannelId(u.shortChannelId.toLong)) match {
+            case Some(pc) =>
+              // The channel was pruned (one of the two channel updates was older than 2 weeks).
+              val pc1 = pc.applyChannelUpdate(update)
+              if (!Announcements.checkSig(u, pc.getNodeIdSameSideAs(u))) {
+                log.warning("bad signature for announcement shortChannelId={} {}", u.shortChannelId, u)
+                sendDecision(origins, GossipDecision.InvalidSignature(u))
+                d
+              } else if (pc1.isStale(currentBlockHeight)) {
+                log.debug("channel shortChannelId={} is still stale and should stay pruned", u.shortChannelId)
+                sendDecision(origins, GossipDecision.RelatedChannelPruned(u))
+                // We update our state with this new channel update.
+                db.updateChannel(u)
+                val prunedChannels1 = d.prunedChannels + (pc1.shortChannelId -> pc1)
+                d.copy(prunedChannels = prunedChannels1)
+              } else {
+                log.info("channel shortChannelId={} is back from the dead", u.shortChannelId)
+                // We notify front nodes that the channel is back.
+                ctx.system.eventStream.publish(ChannelsDiscovered(SingleChannelDiscovered(pc1.ann, pc1.capacity, pc1.update_1_opt, pc1.update_2_opt) :: Nil))
+                sendDecision(origins, GossipDecision.Accepted(u))
+                db.updateChannel(u)
+                // We rebroadcast both updates: both of them are valid now that the channel isn't stale anymore.
```

Comment:
Rebroadcasting both updates is a bit dubious, and order matters in any case: you want to broadcast the most recent first.

Reply:
I think it's necessary if we want to be maximally useful to our peers.
If for example update B -> A is stale.
Whenever we receive an update for the other side (A -> B), we won't rebroadcast them to our peers, so at some point the update they have for A -> B may also become stale.

Once an update for B -> A comes and the channel isn't stale anymore, if we only forward that one to our peers, they may have a stale A -> B update, so they wouldn't restore the channel. We really need to send both to avoid this case.

I don't understand why order would matter? Both channel updates are not stale and for a different channel side, that's all that matters.